### PR TITLE
GH-70: Add Flashlight API

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change ###

Describe your changes here. 

### Bugs Fixed ###

- Related to issue #70 

### API Changes ###

```csharp
static class Flashlight {
    static void On();
    static void Off();
}
```

### Behavioral Changes ###

Adding a new API.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Updated documentation ([see walkthrough](https://github.com/xamarin/Caboodle/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

**Comment by mattleibow** [comment]
Just looking at this Android code, I am wondering if there is something missing. All the other plugins have this scary code with previews and other things:

 - https://github.com/kphillpotts/Xamarin.Plugins/blob/master/Lamp/Lamp/Lamp.Plugin.Android/LampImplementation.cs
 - https://github.com/EddyVerbruggen/Flashlight-PhoneGap-Plugin/blob/master/src/android/nl/xservices/plugins/Flashlight.java

**Comment by jamesmontemagno** [comment]
I think that scary code is needed for < Android M. That is when they added that API

File: Caboodle/Flashlight/Flashlight.android.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            var permission = Manifest.Permission.Camera;
+            if (!Platform.HasPermissionInManifest(permission))
+                throw new PermissionException(permission);
+
+            hasCameraPermission = true;
+        }
+
+        static Task ToggleTorchAsync(bool switchOn)
+        {
+            return Task.Run(() =>
```
Comment (mattleibow) [inline Caboodle/Flashlight/Flashlight.android.cs]: This freezes the UI if not done on a background thread.

File: Caboodle/Flashlight/Flashlight.android.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                        {
+                            var camera = Camera.Open();
+                            if (Platform.HasApiLevel(BuildVersionCodes.Honeycomb))
+                            {
+                                // required for (at least) the Nexus 5
+                                camera.SetPreviewTexture(new SurfaceTexture(0));
+                            }
+                        }
+
+                        var param = camera.GetParameters();
```
Comment (mattleibow) [inline Caboodle/Flashlight/Flashlight.android.cs]: This throws a null ref exception, even though camera is _not_ null ???

File: Caboodle/Flashlight/Flashlight.android.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                {
+                    lock (locker)
+                    {
+                        if (camera == null)
+                        {
+                            var camera = Camera.Open();
+                            if (Platform.HasApiLevel(BuildVersionCodes.Honeycomb))
+                            {
+                                // required for (at least) the Nexus 5
+                                camera.SetPreviewTexture(new SurfaceTexture(0));
```
Comment (mattleibow) [inline Caboodle/Flashlight/Flashlight.android.cs]: Yay for a great OS!

File: Caboodle/Flashlight/Flashlight.android.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            if (!IsSupported)
+                throw new FeatureNotSupportedException();
+
+            await ToggleTorchAsync(false);
+            await ReleaseCameraAsync();
+        }
+
+        static void ValidateCameraPermission()
+        {
+            // TODO: request camera permissions for API Level >= 23
```
Comment (mattleibow) [inline Caboodle/Flashlight/Flashlight.android.cs]: This needs to be looked at for 23+ devices (I don't have any)

File: Caboodle/Flashlight/Flashlight.android.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                        camera.SetParameters(param);
+                        camera.StartPreview();
+                    }
+                }
+            });
+        }
+
+        static async Task ReleaseCameraAsync()
+        {
+            await Task.Run(() =>
```
Comment (mattleibow) [inline Caboodle/Flashlight/Flashlight.android.cs]: This freezes the UI if not done on a background thread.

File: Caboodle/Flashlight/Flashlight.ios.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        static AVCaptureDevice device;
+        static bool hasLoadedDevice;
+
+        public static Task TurnOnAsync()
+        {
+            FindDevice();
+            if (device == null)
+                throw new FeatureNotSupportedException();
+
+            device.SetTorchModeLevel(AVCaptureDevice.MaxAvailableTorchLevel, out var error);
```
Comment (mattleibow) [inline Caboodle/Flashlight/Flashlight.ios.cs]: This just crashes on my device with no stack trace or error message.

File: Caboodle/Flashlight/Flashlight.android.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                        }
+                    }
+                }
+                else
+                {
+                    lock (locker)
+                    {
+                        if (camera == null)
+                        {
+                            var camera = Camera.Open();
```
Comment (Redth) [inline Caboodle/Flashlight/Flashlight.android.cs]: Scope bites again! 😂

**Pull request merged**